### PR TITLE
[WIP] [Draft] Autologging functionality for scikit-learn integration with XGBoost and LightGBM

### LINKED ISSUE #4296: [FR] Autologging functionality for scikit-learn integration with XGBoost (and LightGBM)
## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers. We're seeking help with the implementation of roadmap items tagged with the `help wanted` label. Items tagged with the `needs design` label are awaiting a design sketch from an MLflow maintainer: @WeichenXu123.

For requirements clarifications and implementation questions, or to request a PR review, please tag @WeichenXu123 in your communications related to this issue.

## Proposal Summary

Hello, 

When trying to implement autolog() function for XGBoostClassifier and XGBoostRegressor, we are not able to see the metrics and parameters or models which are getting logged. Can you please address this issue. 

Thanks.

PFA added code for XGBoostRegressor which is to be converted to .py file to replicate the issue
[boston.txt](https://github.com/mlflow/mlflow/files/6386251/boston.txt)

### PR DESCRIPTION
**Pull request by jwyyy**
Signed-off-by: Junwen Yao, jwyiao@gmail.com.

## What changes are proposed in this pull request?

This PR will enable autologging for XGBoost and LightGBM sklearn estimators. Resolves https://github.com/mlflow/mlflow/issues/4296.

(Part 1) autologging for XGBoost sklearn estimators. 
(Part 2) working on autologging for LightGBM sklearn estimators.

## How is this patch tested?

A short example is provided. Tests will be added later.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

This PR will enable autologging for XGBoost (LightGBM) sklearn estimators using `mlflow.xgboost.autolog()` (`mlflow.lightgbm.autolog()`).

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by dbczumar — COMMENTED** [review]
@jwyyy Awesome work! I left some initial feedback about saving / loading / performing inference via pyfunc with XGBoost scikit-learn models. Mainly, I think we should expand `mlflow.xgboost.save_model()` to support XGBoost + scikit-learn without having to call `mlflow.sklearn.save_model()`, as proposed here: https://github.com/mlflow/mlflow/pull/4885/files#r730036638.

For modularity, it may be easier to break this part into a separate PR. Thank you so much for your contributions so far!

**Comment by jwyyy** [comment]
Hi @dbczumar, I studied your suggestions more over the weekend and have more thoughts to share. 

I think we agreed that `mlflow.sklearn` autologging routine is used for logging XGBoost sklearn estimators. This logs model during training using `mlflow.sklearn` flavor (e.g. [`fit_mlflow()`](https://github.com/mlflow/mlflow/blob/master/mlflow/sklearn/__init__.py#L1187-L1200) patched to the original`fit()`). In particular, in `_log_posttraining_metadata()` , it is the `mlflow.sklearn.log_model()` that is called ([L1285](https://github.com/mlflow/mlflow/blob/706758e753dca1a29f11da83d83007058a335fc7/mlflow/sklearn/__init__.py#L1285)). When XGBoost sklearn estimator's `fit()` is patched, the estimator will be saved/logged as a _sklearn estimator_ in mlflow. 

This means that when we load it back, it should be loaded using `mlflow.sklearn` API to make everything consistent with what has been logged during the training. It also avoids `DMatrix` in inference, b/c `predict()` can take inputs without conversion (a `xgboost.sklearn` [implementation](https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/sklearn.py#L829-L904)). 


However, this behavior is different from `mlflow.xgboost.log_model()` and `.save_model()` as you have shown in this [example](https://github.com/mlflow/mlflow/pull/4885/files#r730036638). 

That is, `mlflow.sklearn` saves/logs models in `mlflow.sklearn` flavor, but `mlflow.xgboost` saves/logs models as Boosters. (This also explains why `serialization_format` was added.) _Inconsistency can happen if users try to load model that was previously saved using_ `mlflow.sklearn` _by_ `xgboost.sklearn` _API_.

Originally, the issue (the boston.txt example in https://github.com/mlflow/mlflow/issues/4296) was when XGBooost sklearn estimator is called with `fit()`, the training isn't autologged. (The official sklearn autologging [example](https://github.com/mlflow/mlflow/blob/master/examples/sklearn_autolog/linear_regression.py) logs model automatically.) Explicitly calling `mlflow.xgboost.log_model()` not only logs model as Boosters but also is inconsistent with our proposal: using `mlflow.sklearn` to handle autologging. (Plus, it requires users to manually log training, and there can be early stopping parameters. They need to be checked and logged at each iteration.)

To make it uniform across all cases, we should decide whether
1. saving/loading models are all in `mlflow.sklearn` flavor in both `mlflow.sklearn` and `mlflow.xgboost` for XGBoost sklearn estimators; or 
2. all in `mlflow.xgboost` flavor, i.e., using the provided load_model() and save_model() method in `xgboost.sklearn`.

For (1), we could update `mlflow.xgboost.load_model()` by adding new model flavor `xgboost-sklearn` as you suggested. But I think essentially we still need to utilize functions defined in `mlflow.sklearn`.

Thanks again for your feedback! Looking forward to hearing more discussions from you and @harupy .

**Comment by jwyyy** [comment]
> When `mlflow.sklearn._autolog()` sees that the model class comes from `xgboost`, it should call `mlflow.xgboost.log_model()` instead of `mlflow.sklearn.log_model()`, and `mlflow.xgboost.load_model()` / the XGBoost pyfunc representation should be extended to support XGBoost scikit-learn models.

Hi @dbczumar, thank you for your clarification! Now this internal behavior is clearer to me, and it makes a lot more sense. I will revise the implementation soon. Will let you know if other issues come up.

**Comment by jwyyy** [comment]
Hi @dbczumar @harupy, I revised the implementation as per our discussion and made a new commit. Here is a summary of what have been changed:
1. `mlflow.sklearn` saving / loading are now dropped and directly patched by corresponding saving / loading functions in `mlflow.xgboost`. I didn't use `safe_patch()` since inside sklearn autologging routine, `safe_patch()` will be used to patch `fit()` anyway. Plus, the input arguments are not the same. This resolves the first issue: using `mlflow.xgboost` for both Booster and XGBoost sklearn estimators.
2. Regarding the model class specification, I chose to add a new module level variable `MODEL_CLASS` to `mlflow.xgboost`. This also avoids adding new model specific flavors. An extra flavor key-value, specifying what XGBoost slearn model is used, will be logged in MLmodel: (an example of new MLmodel file)
```
...
flavors:
  python_function:
    data: model.xgb
    env: conda.yaml
    loader_module: mlflow.xgboost
    python_version: 3.6.13
  xgboost:
    data: model.xgb
    model_class: XGBRegressor # new 
    xgb_version: 1.4.2
...
```
(When `xgboost.train()` is used, model_class is Booster.) When the model is loaded, `MODEL_CLASS` will be set to the model class it reads. It resolves the issue of hard saving / loading XGBoost models. Alternatively, we can ask users to specify model class. But I think it is better to have `mlflow` directly handle model class specification. For example, the user who uses a XGBoost model might not be the person who trained it. Manually setting model class requires end users to know which class was used before hand, making the process not very smart.

3. Since `MODEL_CLASS` is new, it should be correctly handled in `mlflow.pyfunc.load_model()`. A few lines were added to set `MODEL_CLASS` before calling module level `_load_pyfunc()`.
4. One drawback of this approach is once `MODEL_CLASS` is set, it should not be used to log other models. So if users want to log another XGBoost model, they need to create a new autologging routine. (I suppose this case is rare also.)

Please let me know if I missed anything. I'd like to hear more feedback and suggestions from you! Thanks!

**Review by dbczumar — COMMENTED** [review]
@jwyyy Awesome progress! Thank you for your ongoing contribution! I've left a few more comments. Can we work on adding support for logging / loading XGBoost scikit-learn models via `mlflow.xgboost` in a separate PR for reviewability purposes?

**Comment by jwyyy** [comment]
Closing this PR, since #4296 is now resolved.
